cmd/kube-controller-manager/app/core.go

```diff
@@ -279,42 +277,33 @@ func startGarbageCollectorController(ctx ControllerContext) (bool, error) {
 		return false, nil
 	}
 
-	// TODO: should use a dynamic RESTMapper built from the discovery results.
-	restMapper := api.Registry.RESTMapper()
-
 	gcClientset := ctx.ClientBuilder.ClientOrDie("generic-garbage-collector")
-	preferredResources, err := gcClientset.Discovery().ServerPreferredResources()
-	if err != nil {
-		return true, fmt.Errorf("failed to get supported resources from server: %v", err)
-	}
-	deletableResources := discovery.FilteredBy(discovery.SupportsAllVerbs{Verbs: []string{"get", "list", "watch", "patch", "update", "delete"}}, preferredResources)
-	deletableGroupVersionResources, err := discovery.GroupVersionResources(deletableResources)
-	if err != nil {
-		return true, fmt.Errorf("Failed to parse resources from server: %v", err)
-	}
+
+	restMapper, syncThread := gcsetup.RESTMapper(gcClientset.Discovery(), 30*time.Second)
```

Comment:
I preferred the construct that got back a caching discovery client and simply started a gofunc that called `reset`.